crates/astria-core/src/primitive/v1/mod.rs

```diff
@@ -490,46 +491,83 @@ where
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord)]
 #[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
 #[cfg_attr(
     feature = "serde",
     serde(try_from = "raw::TransactionId", into = "raw::TransactionId")
 )]
 pub struct TransactionId {
-    hash: String,
+    inner: [u8; TRANSACTION_ID_LEN],
 }
 
 impl TransactionId {
-    /// Returns the hex-encoded 32-byte transaction hash as a string.
+    /// Constructs a new `TransactionId` from a 32-byte array.
     #[must_use]
-    pub fn get(&self) -> String {
-        self.hash.clone()
+    pub const fn new(inner: [u8; TRANSACTION_ID_LEN]) -> Self {
+        Self {
+            inner,
+        }
+    }
+
+    /// Returns the 32-byte transaction hash.
+    #[must_use]
+    pub fn get(self) -> [u8; TRANSACTION_ID_LEN] {
+        self.inner
     }
 
     #[must_use]
     pub fn to_raw(&self) -> raw::TransactionId {
         raw::TransactionId {
-            hash: self.hash.clone(),
+            inner: hex::encode(self.inner),
         }
     }
 
     #[must_use]
     pub fn into_raw(self) -> raw::TransactionId {
         raw::TransactionId {
-            hash: self.hash.clone(),
+            inner: hex::encode(self.inner),
         }
     }
 
-    /// Returns a copy of the transaction hash as a 32-byte array.
+    /// Convert from a raw protobuf type to a rust type for a transaction ID.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the transaction ID buffer was not 32 bytes long or if it was not hex
+    /// encoded.
+    pub fn try_from_raw(raw: raw::TransactionId) -> Result<Self, TransactionIdError> {
+        let inner = <[u8; TRANSACTION_ID_LEN]>::from_hex(raw.inner).map_err(|err| {
+            TransactionIdError(TransactionIdErrorKind::HexDecodeError {
+                source: err,
+            })
+        })?;
+        Ok(Self {
+            inner,
+        })
+    }
+
+    /// Convert from a reference to raw protobuf type to a rust type for a transaction ID.
     ///
-    /// # Panics
+    /// # Errors
     ///
-    /// Panics if the transaction ID is not 32 bytes long. This should never happen, as the
-    /// transaction ID can only be created from a 32-byte array.
+    /// Returns an error if the transaction ID buffer was not 32 bytes long or if it was not hex
+    /// encoded.
+    pub fn try_from_raw_ref(raw: &raw::TransactionId) -> Result<Self, TransactionIdError> {
+        let inner = <[u8; TRANSACTION_ID_LEN]>::from_hex(&raw.inner).map_err(|err| {
+            TransactionIdError(TransactionIdErrorKind::HexDecodeError {
+                source: err,
+            })
+        })?;
+        Ok(Self {
+            inner,
+        })
+    }
+
+    /// Returns a copy of the 32-byte transaction hash as a hex-encoded string.
     #[must_use]
-    pub fn as_bytes(&self) -> [u8; TRANSACTION_ID_LENGTH] {
-        hex::decode(self.hash.clone()).unwrap().try_into().unwrap()
+    pub fn as_hex(&self) -> String {
```

Comment:
Remove. You have already defined `Display`, which does exactly this when you do `to_string()`.